instrumentation/google.golang.org/grpc/otelgrpc/test/interceptor_test.go

```diff
@@ -583,24 +583,67 @@ func TestStreamClientInterceptorWithError(t *testing.T) {
 	assert.Equal(t, codes.Error, span.Status().Code)
 }
 
-func TestServerInterceptorError(t *testing.T) {
-	sr := tracetest.NewSpanRecorder()
-	tp := trace.NewTracerProvider(trace.WithSpanProcessor(sr))
-	usi := otelgrpc.UnaryServerInterceptor(otelgrpc.WithTracerProvider(tp))
-	deniedErr := status.Error(grpc_codes.PermissionDenied, "PERMISSION_DENIED_TEXT")
-	handler := func(_ context.Context, _ interface{}) (interface{}, error) {
-		return nil, deniedErr
+var serverChecks = []struct {
+	grpcCode     grpc_codes.Code
+	wantSpanCode codes.Code
+	name         string
+}{
+	{
+		grpcCode:     grpc_codes.OK,
+		wantSpanCode: codes.Unset,
+		name:         "ok",
+	},
+	{
+		grpcCode:     grpc_codes.Canceled,
+		wantSpanCode: codes.Unset,
+		name:         "cancelled",
+	},
+	{
+		grpcCode:     grpc_codes.NotFound,
+		wantSpanCode: codes.Unset,
+		name:         "not.found",
+	},
+	{
+		grpcCode:     grpc_codes.PermissionDenied,
+		wantSpanCode: codes.Unset,
+		name:         "permission.denied",
+	},
+	{
+		grpcCode:     grpc_codes.Unimplemented,
+		wantSpanCode: codes.Error,
+		name:         "unimplemented",
+	},
+	{
+		grpcCode:     grpc_codes.Internal,
+		wantSpanCode: codes.Error,
+		name:         "internal",
+	},
+	{
+		grpcCode:     grpc_codes.Unauthenticated,
+		wantSpanCode: codes.Unset,
+		name:         "unauthenticated",
+	},
+}
+
+func assertServerErr(t *testing.T, err error, grpcCode grpc_codes.Code) {
+	// the error should be nil only if the gRPC code is OK, otherwise it should be the same as the gRPC error
+	if grpcCode == grpc_codes.OK {
+		assert.NoError(t, err)
+	} else {
+		assert.Equal(t, grpcCode, status.Code(err))
 	}
-	_, err := usi(context.Background(), &grpc_testing.SimpleRequest{}, &grpc.UnaryServerInfo{}, handler)
-	require.Error(t, err)
-	assert.Equal(t, err, deniedErr)
+}
 
-	span, ok := getSpanFromRecorder(sr, "")
-	if !ok {
-		t.Fatalf("failed to export error span")
+func assertServerSpan(t *testing.T, span trace.ReadOnlySpan, wantSpanCode codes.Code, grpcCode grpc_codes.Code, grpcErr error) {
+	// validate span status
+	assert.Equal(t, wantSpanCode, span.Status().Code)
+	if span.Status().Code == codes.Error {
+		assert.Contains(t, grpcErr.Error(), span.Status().Description)
+	} else {
+		assert.Empty(t, span.Status().Description)
```

Comment:
I would prefer to add `wantSpanStatusDescription` to `serverChecks` and change this to:

```suggestion
	assert.Equal(t, wantSpanStatusDescription, span.Status().Description)
```
Having too much conditional in tests logic makes them more obscure, harder to maintain, and less trustworthy.

More on this topic: http://xunitpatterns.com/Conditional%20Test%20Logic.html